### PR TITLE
BUG: Fix for #10533 np.dtype(ctype) does not respect endianness

### PR DESCRIPTION
**Pull request by bmakos**
See: #10533
Added some code that uses the __name__ of the ctypes type. Also
added some new test to make sure we'll know if the ctypes data
that allows this workaround breaks in the future.

### CODE REVIEW

File: numpy/core/_dtype_ctypes.py [hunk 1]
```diff
@@ -61,12 +60,24 @@ def _from_ctypes_structure(t):
     else:
         fields = []
         for fname, ftyp in t._fields_:
-            fields.append((fname, dtype_from_ctypes_type(ftyp)))
+            fields.append((fname, dtype_from_ctypes_endianness(ftyp)))
```
Comment (eric-wieser) [inline numpy/core/_dtype_ctypes.py]: There's no need to change this line or the one above - the recursion will handle it just fine

File: numpy/core/_dtype_ctypes.py [hunk 2]
```diff
@@ -61,12 +60,24 @@ def _from_ctypes_structure(t):
...
 
         # by default, ctypes structs are aligned
         return np.dtype(fields, align=True)
 
 
+def dtype_from_ctypes_endianness(t):
+    """
+    Return the dtype type with endianness included if it's the case
+    """
+    if t.__name__.endswith('_be'):
```
Comment (eric-wieser) [inline numpy/core/_dtype_ctypes.py]: Found something better - `if t.__ctypes_be__ is t`

File: numpy/core/_dtype_ctypes.py [hunk 3]
```diff
@@ -61,12 +60,24 @@ def _from_ctypes_structure(t):
...
         fields = []
         for fname, ftyp in t._fields_:
-            fields.append((fname, dtype_from_ctypes_type(ftyp)))
+            fields.append((fname, dtype_from_ctypes_endianness(ftyp)))
 
         # by default, ctypes structs are aligned
         return np.dtype(fields, align=True)
 
 
+def dtype_from_ctypes_endianness(t):
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: Ok, changed.

File: numpy/core/_dtype_ctypes.py (hunk 1 above)
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: Indeed. Changed.

File: numpy/core/_dtype_ctypes.py (hunk 2 above)
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: I don't follow. 
The ctype type has both `_be` and `_le` in it's `__dict__ `independent of actual endianness.

File: numpy/core/_dtype_ctypes.py [hunk 4]
```diff
@@ -82,6 +93,8 @@ def dtype_from_ctypes_type(t):
         raise NotImplementedError(
             "conversion from ctypes.Union types like {} to dtype"
             .format(t.__name__))
+    elif isinstance(t, type(ctypes.c_int)): # Could be any simple type instead of c_int, all return the same type
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: Done.

[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: Done now, missed it in previous commit.

File: numpy/core/_dtype_ctypes.py [hunk 5]
```diff
@@ -67,6 +66,18 @@ def _from_ctypes_structure(t):
...
+def dtype_from_ctypes_scalar(t):
+    """
+    Return the dtype type with endianness included if it's the case
+    """
+    if t.__ctype_be__ is t:
+        return '>' + t._type_
+    elif t.__ctype_le__ is t:
+        return '<' + t._type_
+    else:
+        return t._type_
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: Yes, did this initially but then changed it for some reason. Anyhow, it indeed works and it's consistent with other def's.

File: numpy/core/_dtype_ctypes.py [hunk 6]
```diff
@@ -82,6 +93,8 @@ def dtype_from_ctypes_type(t):
         raise NotImplementedError(
             "conversion from ctypes.Union types like {} to dtype"
             .format(t.__name__))
+    elif isinstance(t, type(ctypes.c_int)): # Could be any simple type instead of c_int, all return the same type
+        return np.dtype(dtype_from_ctypes_scalar(t))
     elif isinstance(t._type_, str):
         return np.dtype(t._type_)
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/_dtype_ctypes.py]: Done I think.

File: numpy/core/tests/test_dtype.py [hunk 7]
```diff
@@ -856,8 +856,19 @@ class PackedStructure(ctypes.Structure):
             itemsize=18))
         self.check(PackedStructure, expected)
 
-    @pytest.mark.xfail(sys.byteorder != 'little',
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint8.__ctype_be__),('two', ctypes.c_uint32.__ctype_be__)]
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Fixed.

File: numpy/core/tests/test_dtype.py [hunk 8]
```diff
@@ -883,3 +893,29 @@ class PaddedStruct(ctypes.BigEndianStructure):
             ('b', '>H')
         ], align=True)
         self.check(PaddedStruct, expected)
+
+    def test_big_endian_simple_type(self):
+        byte_order = np.dtype(ctypes.c_uint16.__ctype_be__).byteorder
+        if sys.byteorder == 'little':
+           expected = '>'
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Fixed.

File: numpy/core/tests/test_dtype.py [hunk 9]
```diff
@@ -856,8 +856,19 @@ class PackedStructure(ctypes.Structure):
...
+    def test_complex_big_endian_structure_packed(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint8.__ctype_be__),('two', ctypes.c_uint32.__ctype_be__)]
+            _pack_ = 1
+        expected = np.dtype([('one', 'u1'), ('two', '>u4')])
+        self.check(BigEndStruct, expected)
+
+    def test_complex_big_endian_structure(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint16.__ctype_be__),('two', ctypes.c_uint32.__ctype_be__)]
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Removed.

File: numpy/core/tests/test_dtype.py [hunk 10]
```diff
@@ -883,3 +893,29 @@ class PaddedStruct(ctypes.BigEndianStructure):
...
+            expected = '='
+        assert byte_order == expected
+
+    def test_little_endian_simple_type(self):
+        byte_order = np.dtype(ctypes.c_uint16.__ctype_le__).byteorder
+        if sys.byteorder != 'little':
+           expected = '<'
+        else:
+            expected = '='
+        assert byte_order == expected
```
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Merged into 1 test.

[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/core/tests/test_dtype.py]: May as well remove the old test above this comment

File: numpy/core/tests/test_dtype.py [hunk 11]
```diff
@@ -883,3 +886,17 @@ class PaddedStruct(ctypes.BigEndianStructure):
...
+            expected = '>'
+        else:
+            expected = '='
+        assert byte_order == expected
+
+    def test_simple_endian_types(self):
+        assert_equal(np.dtype(ctypes.c_uint16.__ctype_le__), np.dtype('<u2'))
+        assert_equal(np.dtype(ctypes.c_uint16.__ctype_be__), np.dtype('>u2'))
+        assert_equal(np.dtype(ctypes.c_uint8.__ctype_le__), np.dtype('u1'))
+        assert_equal(np.dtype(ctypes.c_uint8.__ctype_be__), np.dtype('u1'))
```
Comment (eric-wieser) [inline numpy/core/tests/test_dtype.py]: You should actually use `self.check` here, it tests both types and instances.

File: numpy/core/tests/test_dtype.py [hunk 12]
```diff
@@ -856,8 +856,13 @@ class PackedStructure(ctypes.Structure):
             itemsize=18))
         self.check(PackedStructure, expected)
 
-    @pytest.mark.xfail(sys.byteorder != 'little',
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
+        class BigEndStruct(ctypes.BigEndianStructure):
+            _fields_ = [('one', ctypes.c_uint8),('two', ctypes.c_uint32)]
```
Comment (eric-wieser) [inline numpy/core/tests/test_dtype.py]: Can you line wrap this like in the tests below?

File: numpy/core/tests/test_dtype.py [hunk 13]
```diff
@@ -856,8 +856,13 @@ class PackedStructure(ctypes.Structure):
             itemsize=18))
         self.check(PackedStructure, expected)
 
-    @pytest.mark.xfail(sys.byteorder != 'little',
-        reason="non-native endianness does not work - see gh-10533")
+    def test_complex_big_endian_structure_packed(self):
```
Comment (eric-wieser) [inline numpy/core/tests/test_dtype.py]: I'd drop the word "complex" - this is no more complex than the test below it.

File: numpy/core/tests/test_dtype.py [hunk 14]
```diff
@@ -856,8 +856,13 @@ class PackedStructure(ctypes.Structure):
...
+        class BigEndStruct(ctypes.BigEndianStructure):
```
Comment (eric-wieser) [inline numpy/core/tests/test_dtype.py]: Can you add a test with a packed LittleEndianStructure too?

File: numpy/core/tests/test_dtype.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Yes, missed to do it. Done now.

File: numpy/core/tests/test_dtype.py (hunk 11 above)
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Done.

File: numpy/core/tests/test_dtype.py (hunk 12 above)
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Yes, done.

File: numpy/core/tests/test_dtype.py (hunk 14 above)
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Added.

File: numpy/core/tests/test_dtype.py (hunk 13 above)
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Done.

File: numpy/core/tests/test_dtype.py (hunk 11 above)
Comment (bmakos) [reply numpy/core/tests/test_dtype.py]: Ok, I just had to look how `check `was defined... Did the commit now.

**Review by eric-wieser — APPROVED** [review]
Looks great, thanks! I plan to squash together all the commits when I merge

**Comment by bmakos** [comment]
Thank you for the review Eric! I enjoyed working on this. I'll go choose some other bug today to check out.

**Comment by eric-wieser** [comment]
If @euler16 doesn't pick it back up, I'm sure you'd have an easy time of #12273. On the one hand, it'd be great to leave them a chance to contribute - but on the other hand, I'd really like to be able to say that 1.16 has full ctypes support. Let's give them a few more days to reply...

**Comment by charris** [comment]
Thanks @bmakos .

**Pull request merged**